dask/array/reductions.py

```diff
@@ -119,6 +119,19 @@ def partial_reduce(func, x, split_every, keepdims=False, dtype=None, name=None):
     return Array(sharedict.merge(x.dask, (name, dsk)), name, out_chunks, dtype=dtype)
 
 
+def empty_type_of(x):
+    return np.ma.empty if isinstance(x, np.ma.masked_array) else np.empty
+
+
+def maybe_ma_divide(x1, x2, dtype=None):
+    if isinstance(x1, np.ma.masked_array) or isinstance(x2, np.ma.masked_array):
+        x = np.ma.divide(x1, x2)
+        if dtype is not None:
+            x = x.astype(dtype)
+        return x
+    return divide(x1, x2, dtype=dtype)
```

Comment:
This function and the one above it (`empty_type_of`) could be easily switched out for a dispatch system, but the changes for `arg` and `cumulative` operations less so. My personal preference is to wait until there's a need for another in memory container, and then figure out what needs to be generalized.